yt/visualization/accumulators.py

```diff
@@ -0,0 +1,290 @@
+    accumulated value of the field is stored at each point along the
+    path, allowing the user to query this information.
+    """
+
+    def __init__(self, paths, ds):
+        self.paths = []
+        self.ds = ds
+        self.ad = ds.all_data()
+        self.accum = []
+        self.left_edge = self.ds.domain_left_edge
+        self.right_edge = self.ds.domain_right_edge
+        # Make sure that the path has proper units. If no units are
+        # specified, assume code units
+        for p in paths:
+            if not isinstance(p, YTArray) or p.units.is_dimensionless:
+                self.paths.append(self.ds.arr(p, "code_length"))
+            else:
+                self.paths.append(p.to(self.ds.length_unit))
+
+    def _get_tree(self, field, is_vector):
+        r"""
+        Creates an AMRKDTree and then adds the field to it.
+
+        Parameters
+        ----------
+        field : list
+            Either the scalar field to add to the tree or an iterable
+            containing the components of the vector field to add.
+
+        is_vector : bool
+            If True, then `field` is a vector and it is assumed that it's
+            an iterable containing the x, y, and z components of the field.
+            Otherwise, the field is assumed to be a scalar field.
+
+        Returns
+        -------
+        tree : AMRKDTree
+            An AMRKDTree for the dataset containing the given field values.
+        """
+        tree = AMRKDTree(self.ds)
+        use_log = [False for i in range(len(field))]
+        field = [self.ad._determine_fields(f)[0] for f in field]
+        tree.set_fields(field, use_log, False)
+        return tree
+
+    def _get_path_field_values(self, tree, path, idx, vals, npts):
+        r"""
+        Determines the value of the field at each point along the path.
+
+        Parameters
+        ----------
+        tree : AMRKDTree
+            An AMRKDTree for the dataset containing the given field values.
+
+        path : np.ndarray
+            The N x 3 array containing the x, y, and z coordinates of the
+            N points used to define the path.
+
+        Returns
+        -------
+        path_field_values : np.ndarray
+            The array containing the field values at each point along the
+            path.
+        """
+        node = tree.locate_node(path[idx])
+        # Data is a list, with one element for each field component
+        data = node.data.my_data
+        # Number of cells in the node
+        dims = data[0].shape
+        # Put the data in a more convenient form: dims x ndims array
+        data = [d.flatten() for d in data]
+        data = ustack(data, axis=1)
+        # Cell width
+        node_left_edge = self.ds.arr(node.get_left_edge(), "code_length")
+        node_right_edge = self.ds.arr(node.get_right_edge(), "code_length")
+        cell_size = (node_right_edge - node_left_edge) / dims
+        while idx < npts:
```

Comment:
I believe this entire loop can be vectorized.

I think what we should do is familiarize ourselves with the behavior -- make sure that we have passing and appropriate tests, that they cover some corner cases, etc, and then turn it into a vectorized loop.  But this definitely can be vectorized, because we're just selecting values here, not computing an integral (until later.)

Reply:
The reason it's not vectorized is because it's recursive, since not every point will be in the same node, and I wasn't sure how to vectorize that.